tracer/src/Datadog.Trace/IAST/Analyzers/HardcodedSecretsAnalyzer.cs

```diff
@@ -0,0 +1,214 @@
+using Datadog.Trace.Logging;
+using Datadog.Trace.Vendors.Serilog.Core;
+
+namespace Datadog.Trace.Iast.Analyzers;
+
+internal class HardcodedSecretsAnalyzer
+{
+    private const int UserStringsArraySize = 100;
+    private static readonly IDatadogLogger Log = DatadogLogging.GetLoggerFor<HardcodedSecretsAnalyzer>();
+    private static HardcodedSecretsAnalyzer? _instance = null;
+
+    private static bool _started = false;
+
+    private static List<SecretRegex>? _secretRules = null;
+
+    private static ManualResetEventSlim _waitEvent = new ManualResetEventSlim(false);
+
+    public HardcodedSecretsAnalyzer()
+    {
+        Log.Debug("HardcodedSecretsAnalyzer -> Init");
+        LifetimeManager.Instance.AddShutdownTask(RunShutdown);
+        _started = true;
+        Task.Run(() => PoolingThread());
+    }
+
+    private static void PoolingThread()
+    {
+        try
+        {
+            Log.Debug("HardcodedSecretsAnalyzer polling thread -> Started");
+            while (_started)
+            {
+                var userStrings = new UserStringInterop[UserStringsArraySize];
+                int userStringLen = NativeMethods.GetUserStrings(userStrings.Length, userStrings);
+                Log.Debug("HardcodedSecretsAnalyzer polling thread -> Retrieved {UserStringLen} strings", userStringLen.ToString());
+                if (userStringLen > 0 && Tracer.Instance.Settings.IsIntegrationEnabled(IntegrationId.HardcodedSecret))
+                {
+                    for (int x = 0; x < userStringLen; x++)
+                    {
+                        var value = Marshal.PtrToStringUni(userStrings[x].Value);
+                        var match = CheckSecret(value!);
+                        if (!string.IsNullOrEmpty(match))
+                        {
+                            var location = Marshal.PtrToStringUni(userStrings[x].Location);
+                            IastModule.OnHardcodedSecret(new Vulnerability(
+                                VulnerabilityTypeName.HardcodedSecret,
+                                (VulnerabilityTypeName.HardcodedSecret + ":" + location!).GetStaticHashCode(),
+                                new Location(location!),
+                                new Evidence(match!),
+                                IntegrationId.HardcodedSecret));
+                        }
+                    }
+
+                    if (userStringLen == userStrings.Length)
+                    {
+                        continue; // Skip wait time if array came full
+                    }
+                }
+
+                _waitEvent.Wait(5_000);
+            }
+        }
+        catch (Exception err)
+        {
+            _started = false;
+            Log.Warning(err, "Exception in HardcodedSecretsAnalyzer polling thread. Disabling feature.");
+        }
+
+        Log.Debug("HardcodedSecretsAnalyzer polling thread -> Exit");
+    }
+
+    internal static string? CheckSecret(string secret)
```

Comment:
I wonder if we should put a try-catch inside here. Currently, if _any_ strings causes an exception with _any_ secret, it will exit the whole loop. That might be a bit excessive.

Also, if we add timeouts to the regex (I think we should) then exceptions may be expected, so we should _definitely_ add a try-catch 😅 I think adding it at the "string" level instead of the "regex" level makes the most sense to me (because if a string crashes one regex, there's a high chance it will crash the others).

WDYT?

Reply:
You are right. On It